### ISSUE TITLE
Coercions with distinct types should traverse pointer modifiers transparently.

### ISSUE THREAD

**Issue by zah** [nim-lang/Nim#7165]
Consider the following code:

``` nim
  type
    Table[K, V] = object
      key: K
      val: V

    MyTable = distinct Table[string, int]
    MyTableRef = ref MyTable

  proc newTable[K, V](): ref Table[K, V] = discard

  proc newMyTable: MyTableRef =
    MyTableRef(newTable[string, int]()) # <--- error here
```

It results in the following error:

```
 Error: type mismatch: got (ref Table[system.string, system.int]) but expected 'MyTableRef = ref MyTable'
```

While it's possible to "cast" the original type to its distinct counterpart directly, when a ref modifier is added, this coercion gets rejected.

The problem seems to be limited to generic types, because a simplified version of the code works correctly:

``` nim
type
  Foo = object
  MyFoo = distinct Foo
  MyFooRef = ref MyFoo

proc newFoo: ref Foo = discard
proc newMyFoo: ref MyFoo =
  MyFooRef(newFoo())
```

A somewhat similar issue was reported before (https://github.com/nim-lang/Nim/issues/6060), but it actually has a small error in the code, because the coercion there is from a ref type to a non-ref type.

**Comment by andreaferretti** [comment]
Is it an issue though? It seems that a `ref distinct T` is not the same as a `distinct ref T`, which makes sense to me

**Comment by zah** [comment]
Well, this is a forced coercion and we tend to allow stripping away the distinct modifiers in this manner.

**Comment by ringabout** [comment]
It has been fixed since Nim 1.0

**Comment by Clyybber** [comment]
Closed with #15368
